pkg/lib/cockpit-components-terminal.jsx

```diff
@@ -120,70 +196,74 @@ export class Terminal extends React.Component {
         this.getText = this.getText.bind(this);
         this.setTerminalTheme = this.setTerminalTheme.bind(this);
 
-        const term = new Term({
-            cols: props.cols || 80,
-            rows: props.rows || 25,
-            screenKeys: true,
-            cursorBlink: true,
-            fontSize: props.fontSize || 16,
-            fontFamily: 'Menlo, Monaco, Consolas, monospace',
-            screenReaderMode: true,
-            showPastingModal: false,
-        });
+        this.terminal_state = this.props.state || new TerminalState();
+        const term = this.terminal_state.terminal;
 
         this.terminalRef = React.createRef();
 
-        term.onData(function(data) {
-            if (this.props.channel.valid) {
-                /* HACK: Ctrl+Space (and possibly other characters) is a disaster: While it is U+00A0 in unicode, with
-                 * an UTF-8 representation of 0xC2A0, the "visible" string in JS is 0x00 (with TextEncoder,
-                 * btoa(), and string comparison). The internal representation retains half of it, and trying to send
-                 * it to the websocket would result in a single 0xA0, which is invalid UTF-8 (and causes the session
-                 * to crash). So intercept and ignore such broken chars.
-                 * See https://github.com/cockpit-project/cockpit/issues/21213 */
-                if (data === '\x00') {
-                    console.log("terminal: ignoring invalid input", data);
-                    return;
-                }
-                this.props.channel.send(data);
-            }
-        }.bind(this));
-
         if (props.onTitleChanged)
             term.onTitleChange(props.onTitleChanged);
 
         this.terminal = term;
         this.state = {
             showPastingModal: false,
-            cols: props.cols || 80,
-            rows: props.rows || 25
+            cols: term.cols,
+            rows: term.rows
         };
     }
 
-    componentDidMount() {
-        this.terminal.open(this.terminalRef.current);
-        this.terminal.loadAddon(new CanvasAddon());
-        this.connectChannel();
+    mountTerminal(state) {
+        this.terminal = state.terminal;
+        this.terminalRef.current.appendChild(state.element);
+        this.terminal.open(state.element);
+        state.connectChannel(this.props.channel);
```

Comment:
And the fact that you do this unconditionally is a bit weird too.  I guess this will be `null` in the case that you are working with an already-existing state, and I'd expect that this would then set the Channel to `null`, but you check for the `null` case in the setter and ignore it...

Reply:
`connectChannel` could move into `Terminal`, actually.